### PR TITLE
#75 destroy hooks

### PR DESCRIPTION
**Pull request by PhilippMDoerner**
This is a branch containing 2 approaches to the entire destructor semantics:
1) (First commit) Use a value object to wrap GdkPixBuf called PixBuf
2) (Second commit) use a type alias to GdkPixBuf called PixBuf directly without any value objects.

To go back to the first attempt, just revert the second commit.

### CODE REVIEW

**Comment by can-lehmann** [comment]
Adding a `=sink` hook fixes it for me. I have no idea why though. I plan to investigate further when I have some more time.

File: owlkettle/gtk.nim [hunk 1]
```diff
@@ -340,6 +340,7 @@ proc g_signal_connect_data*(widget: pointer,
 proc g_type_from_name*(name: cstring): GType
 proc g_object_new*(typ: GType): pointer {.varargs.}
 proc g_object_ref*(obj: pointer)
+proc g_object_weak_ref*(obj: pointer, notify: proc (data, oldObjectPtr: pointer) {.cdecl.}, data: pointer)
```
Comment (PhilippMDoerner) [inline owlkettle/gtk.nim:343]: I feel tempted to leave this there just to have it as it shouldn't hurt, disagreement here?

**Comment by PhilippMDoerner** [comment]
We're running into 
```txt
Error: 403 - {"message":"API rate limit exceeded for 13.88.98.81. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
```
The tests themselves *should* run, at least they work for me on 2.0 locally.
However I also modified the path used.

File: owlkettle/gtk.nim (hunk 1 above)
Comment (can-lehmann) [reply owlkettle/gtk.nim:343]: Yes, I would also leave it. It might be useful for debugging in the future.

File: owlkettle/customPragmas.nim [hunk 2]
```diff
@@ -0,0 +1,8 @@
+## Definess a custom pragma called "locker".
+## This does nothing when compiled with a nim version 2.0 or higher,
+## but applies the "locks: 0" pragma if compiled for e.g. 1.6.X.
+## "locks: 0" is applied to ensure that this code never accesses locked data in user-defined procs.
+when NimMajor >= 2:
+  {.pragma: locker.}
+else:
+  {.pragma: locker, locks: 0.}
```
Comment (can-lehmann) [inline owlkettle/customPragmas.nim:8]: I think this would nicely fit into `common.nim`

**Comment by can-lehmann** [comment]
I think we can move forward with 3. for now. You could use a `template` to reduce code duplication (put that into `common.nim` as well).

File: owlkettle/customPragmas.nim (hunk 2 above)
Comment (PhilippMDoerner) [reply owlkettle/customPragmas.nim:8]: Based on your lower comment I assume you want to move this into a `customPragma` template and use that in the various modules?

Comment (can-lehmann) [reply owlkettle/customPragmas.nim]: Oh, sorry, no, I meant putting the `=destroy` in a template for supporting both 1.6.x and 2.0.0. But if a template is need here, using one is also fine.

**Comment by PhilippMDoerner** [comment]
> I think we can move forward with 3. for now. You could use a `template` to reduce code duplication (put that into `common.nim` as well).

I'm actually somewhat torn here.
Because as far as I can see judge freeing the memory for a Pixbuf is a different task than freeing the memory of a TextBuffer is a different task than freeing the memory of a Cstring.
At least for now that's what it looks like to me.
Given that they don't seem inherently coupled to me and it just so happens that their destructors look similar, I hesitate to start coupling some of them together (because cstrings will most likely need their own strategy) when they should be different things.

Feel free to overrule me here, that's just my current instinct from my very limited experience with owlkettle and GTK.

**Comment by can-lehmann** [comment]
I agree, I would do something like:

```nim
template destructor(name, typ, body: untyped) =
  when(NimMajor >= 2):
    proc `=destroy`(name: typ) =
      body
  else:
    proc `=destroy`(name: var typ) =
      body

destructor(pixbuf, Pixbuf):
  ...
```

File: owlkettle/customPragmas.nim
Comment (can-lehmann) [inline owlkettle/customPragmas.nim:1]: This file needs to be deleted from git.

File: owlkettle/widgets.nim [hunk 3]
```diff
@@ -1995,11 +2032,15 @@ type
...
+when(NimMajor >= 2):
+  proc `=destroy`(buffer: TextBufferObj) =
+    g_object_unref(pointer(buffer.gtk))
+else:
+  proc `=destroy`(buffer: var TextBufferObj) =
+    g_object_unref(pointer(buffer.gtk))
 
 proc newTextBuffer*(): TextBuffer =
-  new(result, finalizer=finalizer)
+  new(result)
```
Comment (can-lehmann) [inline owlkettle/widgets.nim:2043]:
```suggestion
  result = TextBuffer(gtk: gtk_text_buffer_new(nil.GtkTextTagTable))
```

File: owlkettle/widgets.nim [hunk 4]
```diff
@@ -1995,11 +2032,15 @@ type
...
   result.gtk = gtk_text_buffer_new(nil.GtkTextTagTable)
```
Comment (can-lehmann) [inline owlkettle/widgets.nim:2044]:
```suggestion

```

File: owlkettle/widgets.nim [hunk 5]
```diff
@@ -458,18 +459,48 @@ type
...
+  if not isNil(source.gdk):
+    g_object_ref(pointer(source.gdk))
+    
+  dest.gdk = source.gdk
+    
 proc newPixbuf(gdk: GdkPixbuf): Pixbuf =
   if gdk.isNil:
     raise newException(ValueError, "Unable to create Pixbuf from GdkPixbuf(nil)")
-  new(result, finalizer=finalizer)
+  result = Pixbuf()
```
Comment (can-lehmann) [inline owlkettle/widgets.nim:501]:
```suggestion
  result = Pixbuf(gdk: gdk)
```

File: owlkettle/widgets.nim [hunk 6]
```diff
@@ -458,18 +459,48 @@ type
...
   result.gdk = gdk
```
Comment (can-lehmann) [inline owlkettle/widgets.nim:502]:
```suggestion

```

**Comment by PhilippMDoerner** [comment]
Applied the template to TextBuf, PixBuf and Stylesheet and streamlined the two creation procs for textbuf and pixbuf as suggested.

**Comment by can-lehmann** [comment]
Are the `=copy` hooks still required? They are currently applied inconsistently.

File: owlkettle/mainloop.nim [hunk 7]
```diff
@@ -20,11 +20,32 @@
...
+    provider: GtkCssProvider
+    priority: int
+
+  Stylesheet* = ref StylesheetObj
+
+crossVersionDestructor(stylesheet, StylesheetObj):
+    if isNil(stylesheet.provider):
+      return
+    
+    g_object_unref(pointer(stylesheet.provider))
```
Comment (can-lehmann) [inline owlkettle/mainloop.nim:36]:
```suggestion
  if isNil(stylesheet.provider):
    return
  
  g_object_unref(pointer(stylesheet.provider))
```

**Comment by PhilippMDoerner** [comment]
> Are the `=copy` hooks still required? They are currently applied inconsistently.

Logically speaking they *should* be required, because a copy means the ref count for gtk goes up by one and they must be informed of that so that they can handle memory management.
I swear I wrote a copy-hook for textbuf at one point, must've gotten deleted over the course of the PR.

**Comment by can-lehmann** [comment]
Since it is a `ref object` though, there should not be any copys (only if you do `pixbuf1[] = pixbuf2[]`).

**Comment by PhilippMDoerner** [comment]
Which is a fair point but in that case I'd still want to have the copy-hook, just for "correctness" even if nobody should ever do that.
So yeah, give me a sec I'll provide a copy hook

File: owlkettle.nimble [hunk 8]
```diff
@@ -18,7 +18,7 @@ task examples, "Build examples":
   withDir "examples":
     for file in findExamples("."):
       echo "INFO: Compile " & file
-      exec "nim c --hints:off --verbosity:0 " & file
+      exec "nim c --hints:off --path:.. --verbosity:0 " & file
```
Comment (can-lehmann) [inline owlkettle.nimble:21]: Why is this required?

Comment (PhilippMDoerner) [reply owlkettle.nimble:21]: I currently very intentionally have no global version of owlkettle installed to make sure when compiling this it **has** to use the local version. If I don't provide a path to this local version (aka the owlkettle **in this repo** instead of the installed owlkettle version) I get:
```nim
/home/philipp/dev/owlkettle/examples/introduction.nim(23, 8) Error: cannot open file: owlkettle
```

Which indicates to me that if you run `nimble examples` as "testing" locally, it'll not use your local owlkettle version.
--path:.. allows you to specify that it shall add wherever your owlkettle repo is to the path, where it can find and thus import owlkettle.nim

This **should** guarantee that it uses the owlkettle files of your repository rather than of whatever owlkettle version you got installed.

Comment (can-lehmann) [reply owlkettle.nimble:21]: Ah, makes sense.

**Comment by can-lehmann** [comment]
Looks good!
